### DISCUSSION TITLE
Custom installation directory?

### DISCUSSION THREAD

**Discussion by adnelson** [keplerproject/luarocks#1602] [Q&A]
I'm trying to build packages with luarocks using the nix package manager, which means that I need to be able to have luarocks install a package into a prefix path that will be generated automatically. I'd love a situation in which I could both:
- Specify any number of paths in which luarocks can find existing dependencies (rather than just `$HOME/.luarocks` and global paths), and
- Specify a custom output directory (again, rather than just `$HOME/.luarocks` and global paths).

Peeking around through the documentation, I haven't seen any way to accomplish this. Although rockspec files can refer to a `PREFIX` variable, but I'm not clear on how this is set. I would love to be able to do something like:

```
luarocks make my-package.rockspec --prefix=/path/to/custom/prefix
```

And have all of the generated binaries and .lua files be output to the given prefix. Similarly, passing in the search paths to dependencies and their metadata would be necessary. Of course, using an environment variable rather than a command-line flag would also be fine.

**Comment by daurnimator** [comment]
You want `--tree`

**Comment by adnelson** [comment]
@daurnimator `--tree` works to set the output directory, thanks! Unfortunately it doesn't seem to work on the input directories. I can't seem to get luarocks to pick up the packages by writing a configfile either. For example, I have the package `luafilesystem` installed in the folder `/nix/store/9gmxg8pfkf103rmzya54w83i52v0n436-lua5.1-filesystem-1.6.3-1`. Here are the files installed in that folder:

``` bash
$ (cd /nix/store/9gmxg8pfkf103rmzya54w83i52v0n436-lua5.1-filesystem-1.6.3-1; find .)
.
./lib
./lib/luarocks
./lib/luarocks/rocks
./lib/luarocks/rocks/luafilesystem
./lib/luarocks/rocks/luafilesystem/1.6.3-1
./lib/luarocks/rocks/luafilesystem/1.6.3-1/doc
./lib/luarocks/rocks/luafilesystem/1.6.3-1/doc/us
./lib/luarocks/rocks/luafilesystem/1.6.3-1/doc/us/doc.css
./lib/luarocks/rocks/luafilesystem/1.6.3-1/doc/us/examples.html
./lib/luarocks/rocks/luafilesystem/1.6.3-1/doc/us/index.html
./lib/luarocks/rocks/luafilesystem/1.6.3-1/doc/us/license.html
./lib/luarocks/rocks/luafilesystem/1.6.3-1/doc/us/luafilesystem.png
./lib/luarocks/rocks/luafilesystem/1.6.3-1/doc/us/manual.html
./lib/luarocks/rocks/luafilesystem/1.6.3-1/tests
./lib/luarocks/rocks/luafilesystem/1.6.3-1/tests/test.lua
./lib/luarocks/rocks/luafilesystem/1.6.3-1/luafilesystem-1.6.3-1.rockspec
./lib/luarocks/rocks/luafilesystem/1.6.3-1/rock_manifest
./lib/luarocks/rocks/manifest
./lib/lua
./lib/lua/5.1
./lib/lua/5.1/lfs.so
```

So it looks like all of the files are there. But now I'm trying to install `lub`, which depends on `luafilesystem >= 1.4.0`. I created a configfile which looks like

``` lua
-- config.lua
rocks_trees = {
  -- Path to installed luafilesystem module
  "/nix/store/9gmxg8pfkf103rmzya54w83i52v0n436-lua5.1-filesystem-1.6.3-1",
  -- Path into which to install lub
  "/nix/store/x2sz9kiv9pqvdx5n3cs12jm8c9k9s527-lua5.1-lub-1.1.0-1"
}
```

And then I'm passing in that config file path calling

```
$ LUAROCKS_CONFIG=config.lua luarocks install lub-1.1.0-1.rockspec
```

But unfortunately, luarocks doesn't pick up the installed `luafilesystem`:

```
Missing dependencies for lub:
luafilesystem >= 1.4.0
```

What do I need to do to get luarocks to discover the installed packages?

**Comment by daurnimator** [comment]
If you already know you have dependencies installed (i.e. you're tracking them elsewhere), use the `--dep-mode=none` flag to `install`.

To actually then use the libraries, you can use `luarocks path --tree <tree> --append` for each library; or use a custom loader. (perhaps there is a way to use the custom loader shipped with luarocks for this; but I'm not sure)

**Comment by rnc** [comment]
@daurnimator  I can't see `--tree` documented in https://github.com/keplerproject/luarocks/wiki/install or in the `--help` flag.
Is there a way of controlling whether it installs to lib or lib64 directory?

**Comment by daurnimator** [comment]
> @daurnimator I can't see --tree documented in https://github.com/keplerproject/luarocks/wiki/install or in the --help flag.

Look in plain old `luarocks --help`, not in the `install` subcommand. Or see it in https://github.com/keplerproject/luarocks/wiki/luarocks

**Comment by rnc** [comment]
Ah ok. Can --tree work with every subcommand?

**Comment by hishamhm** [comment]
@rnc yes!
